packages/esm-patient-chart-app/src/patient-chart/patient-chart.component.tsx

```diff
@@ -16,31 +16,85 @@ import ChartReview from '../patient-chart/chart-review/chart-review.component';
 import { type LayoutMode } from './chart-review/dashboard-view.component';
 import styles from './patient-chart.scss';
+import { useVisitByUuId } from './patient-chart.resources';
 
 const PatientChart: React.FC = () => {
   const { patientUuid, view: encodedView } = useParams();
   const view = decodeURIComponent(encodedView);
   const { isLoading: isLoadingPatient, patient } = usePatient(patientUuid);
-  const state = useMemo(() => ({ patient, patientUuid }), [patient, patientUuid]);
   const { workspaceWindowState, active } = useWorkspaces();
   const [layoutMode, setLayoutMode] = useState<LayoutMode>();
-  // Keep state updated with the current patient. Anything used outside the patient
-  // chart (e.g., the current visit is used by the Active Visit Tag used in the
-  // patient search) must be updated in the callback, which is called when the patient
-  // chart unmounts.
+  const {
+    patientUuid: storePatientUuid,
+    setPatient,
+    visitContext,
+    mutateVisitContext,
+    setVisitContext,
+  } = usePatientChartStore(patientUuid);
+
+  // The patient chart store sets the patient when we enter the patient chart
+  // and unsets the patient when we leave. (This gives extensions and workspaces a way
+  // to check whether they are rendered within the patient chart app.)
+  // Note, however, that the visitContext is NOT unset when we leave the chart, as
+  // we want to persist it across refreshes.
+  // When we enter the chart, we want to update the visit context as follows:
+  // does the the stored visitContext exist and belong to the patient?
+  // 1. If so, the visitContext should be valid but possibly stale; fetch the visit again
+  //    and update the context
+  // 2. If not, fetch the active visit of the patient, If it exists, set it as the
+  //    visitContext; otherwise, clear it.
+  const isVisitContextValid = visitContext && visitContext.patient.uuid === patientUuid;
+  const {
+    data: newVisitContext,
+    mutate: newMutateVisitContext,
+    isValidating: isValidatingVisitContext,
+  } = useVisitByUuId(isVisitContextValid ? visitContext.uuid : null);
+  const {
+    activeVisit,
+    isValidating: isValidatingActiveVisit,
+    mutate: mutateActiveVisit,
+  } = useVisit(isVisitContextValid ? null : patientUuid);
+
+  const state = useMemo(
+    () => ({
+      patientUuid,
+      patient: patient ?? {},
+      visitContext,
+      mutateVisitContext,
+    }),
+    [patient, patientUuid, visitContext, mutateVisitContext],
+  );
+
   useEffect(() => {
-    setCurrentVisit(patientUuid, null);
-    return () => {
-      setCurrentVisit(null, null);
-    };
-  }, [patientUuid]);
+    if (!isValidatingVisitContext && !isValidatingActiveVisit && storePatientUuid) {
+      if (activeVisit) {
+        setVisitContext(activeVisit, mutateActiveVisit);
+      } else if (newVisitContext) {
+        setVisitContext(newVisitContext.data, newMutateVisitContext);
+      } else {
+        setVisitContext(null, null);
+      }
+    }
```

Comment:
Does it maybe make sense to refactor all of this into it's own hook rather than doing it as part of this component?